Added the test to reproduce the data race condition during ledger processing blocks.This test creates one goroutine that repeatedly writes ls.currentEra under ls.Lock(), simulating ledgerProcessBlocks updating the ledger era. At the same time, multiple goroutines repeatedly call ls.calculateStabilityWindow(), simulating chainsync reading the stability window. The bug is that calculateStabilityWindow() reads ls.currentEra.Id without ls.RLock(), while another goroutine may be writing ls.currentEra. With this test, we can confirm that calculateStabilityWindow races with concurrent currentEra updates.

Normal go test passes because no assertion fails here in which go test -race fails because it detects unsafe concurrent read/write.

```
akhil-mac@Akhils-MacBook-Air dingo % 
akhil-mac@Akhils-MacBook-Air dingo % go test -race ./ledger -run TestCalculateStabilityWindowConcurrentCurrentEraAccess -count=1
# github.com/blinklabs-io/dingo/ledger.test
ld: warning: '/private/var/folders/nd/qcslwcz912l131j36zy8c3k00000gn/T/go-link-1804145038/000017.o' has malformed LC_DYSYMTAB, expected 98 undefined symbols to start at index 1626, found 95 undefined symbols starting at index 1626
==================
WARNING: DATA RACE
Write at 0x00c000b0c148 by goroutine 20:
  github.com/blinklabs-io/dingo/ledger.TestCalculateStabilityWindowConcurrentCurrentEraAccess.func1()
      /Users/akhil-mac/1371/dingo/ledger/state_test.go:94 +0xf4

Previous read at 0x00c000b0c148 by goroutine 26:
  github.com/blinklabs-io/dingo/ledger.(*LedgerState).calculateStabilityWindow()
      /Users/akhil-mac/1371/dingo/ledger/state.go:1975 +0xa4
  github.com/blinklabs-io/dingo/ledger.TestCalculateStabilityWindowConcurrentCurrentEraAccess.func2()
      /Users/akhil-mac/1371/dingo/ledger/state_test.go:113 +0xa0

Goroutine 20 (running) created at:
  github.com/blinklabs-io/dingo/ledger.TestCalculateStabilityWindowConcurrentCurrentEraAccess()
      /Users/akhil-mac/1371/dingo/ledger/state_test.go:88 +0x4c4
  testing.tRunner()
      /Users/akhil-mac/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.8.darwin-arm64/src/testing/testing.go:1934 +0x164
  testing.(*T).Run.gowrap1()
      /Users/akhil-mac/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.8.darwin-arm64/src/testing/testing.go:1997 +0x3c

Goroutine 26 (running) created at:
  github.com/blinklabs-io/dingo/ledger.TestCalculateStabilityWindowConcurrentCurrentEraAccess()
      /Users/akhil-mac/1371/dingo/ledger/state_test.go:105 +0x4d4
  testing.tRunner()
      /Users/akhil-mac/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.8.darwin-arm64/src/testing/testing.go:1934 +0x164
  testing.(*T).Run.gowrap1()
      /Users/akhil-mac/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.8.darwin-arm64/src/testing/testing.go:1997 +0x3c
==================
==================
WARNING: DATA RACE
Write at 0x00c000b0c148 by goroutine 20:
  github.com/blinklabs-io/dingo/ledger.TestCalculateStabilityWindowConcurrentCurrentEraAccess.func1()
      /Users/akhil-mac/1371/dingo/ledger/state_test.go:96 +0x160

Previous read at 0x00c000b0c148 by goroutine 25:
  github.com/blinklabs-io/dingo/ledger.(*LedgerState).calculateStabilityWindow()
      /Users/akhil-mac/1371/dingo/ledger/state.go:1975 +0xa4
  github.com/blinklabs-io/dingo/ledger.TestCalculateStabilityWindowConcurrentCurrentEraAccess.func2()
      /Users/akhil-mac/1371/dingo/ledger/state_test.go:113 +0xa0

Goroutine 20 (running) created at:
  github.com/blinklabs-io/dingo/ledger.TestCalculateStabilityWindowConcurrentCurrentEraAccess()
      /Users/akhil-mac/1371/dingo/ledger/state_test.go:88 +0x4c4
  testing.tRunner()
      /Users/akhil-mac/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.8.darwin-arm64/src/testing/testing.go:1934 +0x164
  testing.(*T).Run.gowrap1()
      /Users/akhil-mac/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.8.darwin-arm64/src/testing/testing.go:1997 +0x3c

Goroutine 25 (running) created at:
  github.com/blinklabs-io/dingo/ledger.TestCalculateStabilityWindowConcurrentCurrentEraAccess()
      /Users/akhil-mac/1371/dingo/ledger/state_test.go:105 +0x4d4
  testing.tRunner()
      /Users/akhil-mac/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.8.darwin-arm64/src/testing/testing.go:1934 +0x164
  testing.(*T).Run.gowrap1()
      /Users/akhil-mac/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.25.8.darwin-arm64/src/testing/testing.go:1997 +0x3c
==================
--- FAIL: TestCalculateStabilityWindowConcurrentCurrentEraAccess (0.00s)
    testing.go:1617: race detected during execution of test
FAIL
FAIL    github.com/blinklabs-io/dingo/ledger    0.982s
FAIL
akhil-mac@Akhils-MacBook-Air dingo % 
```

Failure from go-test workflow before making fix:
<img width="1159" height="971" alt="Screenshot 2026-05-06 at 2 03 40 PM" src="https://github.com/user-attachments/assets/f4079a69-4848-4ef7-9b0b-ed8faa05702f" />

--- a/ledger/state.go
+++ b/ledger/state.go
@@ -1972,7 +1972,10 @@ func (ls *LedgerState) IsAtTip() bool {
 // For Byron era, returns 2k. For Shelley+ eras, returns 3k/f.
 // Returns the default threshold if genesis data is unavailable or invalid.
 func (ls *LedgerState) calculateStabilityWindow() uint64 {
-	return ls.calculateStabilityWindowForEra(ls.currentEra.Id)
+	ls.RLock()
+	eraId := ls.currentEra.Id
+	ls.RUnlock()
+	return ls.calculateStabilityWindowForEra(eraId)
 }
 
 // calculateStabilityWindowForEra calculates the stability window for the given era.
@@ -2126,8 +2129,6 @@ func (ls *LedgerState) SecurityParam() int {
 // current era in slots. For Byron the window is 2k; for Shelley+ it is 3k/f.
 // It is safe to call from multiple goroutines.
 func (ls *LedgerState) StabilityWindow() uint64 {
-	ls.RLock()
-	defer ls.RUnlock()
 	return ls.calculateStabilityWindow()
 }
 

--- a/ledger/state_test.go
+++ b/ledger/state_test.go
@@ -68,10 +68,8 @@ func TestLedgerProcessBlocksFromSourceReturnsNilWhenReaderCloses(
 	require.NoError(t, err)
 }
 
-// It reproduces the race where block processing updates currentEra while
-// calculateStabilityWindow reads currentEra.Id without taking ls.RLock().
-// This test is expected to fail under go test -race until the reader is
-// synchronized with the writer.
+// It verifies that calculating the stability window is synchronized with
+// concurrent currentEra updates from block processing.
 func TestCalculateStabilityWindowConcurrentCurrentEraAccess(t *testing.T) {
 	ls := &LedgerState{
 		currentEra: eras.ShelleyEraDesc,

